File: app/src/test/java/com/photoviewer/domain/usecases/GetPhotosListTest.java
package com.photoviewer.domain.usecases;

import com.photoviewer.data.entity.PhotoEntity;
import com.photoviewer.data.repository.PhotoEntityDataSource;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import rx.Observable;
import rx.Scheduler;

import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;

public class GetPhotosListTest {

    private GetPhotosList mGetPhotosList;

    @Mock
    private PhotoEntityDataSource mMockPhotoEntityDataSource;
    @Mock
    private Scheduler mMockScheduler;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        mGetPhotosList = new GetPhotosList(mMockScheduler,
                mMockScheduler, mMockPhotoEntityDataSource);
    }

    @Test
    public void testGetPhotosList() {
        List<PhotoEntity> photos = new ArrayList<PhotoEntity>() {{
            add(new PhotoEntity());
        }};

        given(mMockPhotoEntityDataSource.photos()).willReturn(Observable.just(photos));

        mGetPhotosList.buildObservable();

        verify(mMockPhotoEntityDataSource).photos();
        verifyZeroInteractions(mMockScheduler);
    }
}

File: app/src/test/java/com/photoviewer/domain/usecases/GetPhotosListTest.kt
@file:Suppress("IllegalIdentifier")

package com.photoviewer.domain.usecases

import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import com.photoviewer.data.entity.PhotoEntity
import com.photoviewer.data.repository.PhotoEntityDataSource
import com.photoviewer.domain.Photo
import org.junit.Before
import org.junit.Test
import rx.Observable.just
import rx.Scheduler
import rx.observers.TestSubscriber
import java.util.*
import kotlin.test.assertEquals

class GetPhotosListTest {

    private val mPhotoEntityDataSource: PhotoEntityDataSource = mock()
    private val mMockScheduler: Scheduler = mock()

    private lateinit var mGetPhotosList: GetPhotosList
    private lateinit var mTestSubscriber: TestSubscriber<List<Photo>>

    @Before fun setUp() {
        mTestSubscriber = TestSubscriber.create()
        mGetPhotosList = GetPhotosList(mMockScheduler, mMockScheduler, mPhotoEntityDataSource)
    }

    @Test fun `should get particular photo`() {
        val photos = createPhotosList()

        assumeDataSourceHasRequestedPhotos(photos)

        mGetPhotosList.call().subscribe(mTestSubscriber)

        assertEquals(photos.size, mTestSubscriber.onNextEvents[0].size)
    }

    private fun assumeDataSourceHasRequestedPhotos(photos: ArrayList<PhotoEntity>) {
        whenever(mPhotoEntityDataSource.photos()).thenReturn(just<List<PhotoEntity>>(photos))
    }

    private fun createPhotosList() = ArrayList<PhotoEntity>().apply {
        add(PhotoEntity())
    }
}


